torch/distributed/_composable/fsdp/_fsdp_init.py

```diff
@@ -81,3 +82,32 @@ def _get_managed_states(
             buffers.append(buffer)
             visited_buffers.add(buffer)
     return params, buffers
+
+
+def _move_states_to_device(
+    params: List[nn.Parameter],
+    buffers: List[torch.Tensor],
+    device: torch.device,
+    mesh_info: FSDPMeshInfo,
+) -> None:
+    """
+    We have FSDP move states to device for simpler and faster initialization
+    since FSDP almost always uses CUDA for training. We move parameters/buffers
+    rather than modules since modules to support ignoring parameters/buffers in
+    the future.
+    """
+    # TODO: De-duplicate with `_apply` after `swap_tensors` path lands:
+    # https://github.com/pytorch/pytorch/issues/115792
+    for tensor in itertools.chain(params, buffers):
+        if tensor.device == device:
+            continue
+        if isinstance(tensor, DTensor):
+            if (dtensor_mesh_type := tensor._spec.mesh.device_type) != device.type:
+                raise ValueError(
+                    "Requires DTensor to have mesh of the same type as the FSDP mesh "
+                    f"but got {dtensor_mesh_type} for DTensor and {device.type} for FSDP"
+                )
+            raise AssertionError(
```

Comment:
this assertion error seems always happening if it's a DTensor and it does not hit the previous ValueError?

Reply:
We are expecting that `DTensor` has already moved the tensor to the device. If that happened, then `if tensor.device == device: continue` runs. This assertion is just confirming that the `DTensor` behavior is what we assumed.